### DISCUSSION TITLE
Export invoices with plugin

### DISCUSSION THREAD

**Discussion by Scriptkiddi** [kimai/kimai#2630] [Questions & Answers]
Hey so I'm currently working on an invoice exporter plugin, so we can integrate it into our usual billing system.
My current function is hooked on `InvoicePostRenderEvent` because I need the positions that go into the invoice.
That brings a couple problems, like a preview also leads to an export. 
I would prefer hooking InvoiceCreatedEvent, but from the InvoiceCreatedEvent I can not access the invoice positions.
So as far as I can see I can either add an attribute in the invoice object called positions or something like that or pass the model into the InvoiceCreatedEvent.
My question being which way would you like to see in a pr @kevinpapst or is there an alternative I overlooked? 
Or is it something that does not fit into your development plans and I have to find something completely different to export the invoices?

**Comment by kevinpapst** [comment]
Multiple possible solutions:

1. You can use an InvoiceRenderer to export the invoice directly. Do not store it in Kimai, but directly transfer the data. 
The existing InvoiceRenderer creates PDF/HTML... files, but why not calling the your API?

2. Help me with https://github.com/kevinpapst/kimai2/pull/2414 than you can get the "invoice positions" directly from the Invoice entity.

3. Add the Invoice Model as attribute to the `InvoiceCreatedEvent`.

I would prefer solution **2**.

**Reply by Scriptkiddi** [reply]
I would be willing to help, so link entities means bidirectional? 
The steps would be:

Invoice -> Entry
- Adding a entries attr on the invoice model
- adding a getEntries function to an invoice model
- ?saving entries in the setModel function?

Entry -> Invoice
- Add invoice attr on the ?Timesheet? model
- Activate and implement the setInvoice function in the InvoiceItemInterface

Did I miss anything?

**Reply by kevinpapst** [reply]
No, not bidirectional. Expenses for example can also be included in the invoices. So a bidirectional way would be wrong (or at least misleading) here.

No changes allowed in the interfaces (only new commented method definitions, otherwise this would be a hard BC!). 
We need `method_exists()` calls where you want to use the new methods (including @deprecated annotation) ...  but I can help with that!

There should be 
- a new (commented) method `getInvoiceItemsByInvoice()` in [InvoiceItemRepositoryInterface](https://github.com/kevinpapst/kimai2/blob/master/src/Invoice/InvoiceItemRepositoryInterface.php)
- a new (commented) method `markItemsAsInvoiced(Invoice $invoice, array $items)` in [InvoiceItemRepositoryInterface](https://github.com/kevinpapst/kimai2/blob/master/src/Invoice/InvoiceItemRepositoryInterface.php) 
- the new implementation of this method in [TimesheetInvoiceItemRepository](https://github.com/kevinpapst/kimai2/blob/master/src/Repository/TimesheetInvoiceItemRepository.php)
- After [this line](https://github.com/kevinpapst/kimai2/blob/master/src/Invoice/ServiceInvoice.php#L379) we have to add some lines of code:
```php
        foreach ($this->getInvoiceItemRepositories() as $repository) {
            if (!method_exists('markItemsAsInvoiced', $repository) {
               continue;
            }
            $repository->markItemsAsInvoiced($invoice, $model->getEntries());
        }
```

That should be it more or less if you branch from [entity-to-invoice](https://github.com/kevinpapst/kimai2/tree/entity-to-invoice).

**Reply by Conni2461** [reply]
Hey @kevinpapst,

I am supose to finish this and just wanted to ask if this information is still up to date.

Thanks

**Reply by kevinpapst** [reply]
I hope so. I haven't worked in that topic since I posted this info.

**Comment by kevinpapst** [comment]
Sorry that this took ages to create. I splitted my plan into two steps. 
First one is now in master: https://github.com/kevinpapst/kimai2/pull/3079 
Now you can access the InvoiceModel from the InvoiceCreatedEvent.

The next step is to link timesheet to invoices, which is done in #3080 so they can be accessed later on.

**Comment by labor4** [comment]
Hi
Sound promising. 
Do you still plan to do this, or has this been implemented already in another way? (ie paid plugin)

I would love to POST/export to another system, but my hands are bound with this framework. Way too high for my skills. 🙃

Alternative:
I have had good experiences with the kimai API, and it only occurs to me now, I could "remote-control" an export from another app, as in "readout, and set to exported". Takes a bit of work, but I know I can do this. 
I would value your opinion, if this approach is "API-feature-complete" (including expenses).

Regards

**Reply by kevinpapst** [reply]
I wrote exporter for customer to other CRMs. Nothing that works as plugin out of the box though, they are custom tailored to the customers needs.

Which software do you use for invoicing?

**Reply by labor4** [reply]
Hi Kevin
My own.
I would need a json POST from the Kimai invoice/exporter panel to an external HTTP site.
The parsing/storing I sort out myself there.
- rows with full entries
- plus maybe a misc-array with overall specs that are not repeating

**Reply by kevinpapst** [reply]
Easy: build a plugin with an EventSubscriber (copy & paste from DemoBundle), fetch all entries from the InvoiceModel and send the POST

**Reply by Scriptkiddi** [reply]
little bit late to the conversation but we are doing something like this here https://git.helsinki.tools/helsinki-systems/LexofficeBundle

**Reply by kevinpapst** [reply]
@Scriptkiddi you should really post that to the marketplace: https://www.kimai.org/documentation/store.html
